core/src/com/etheller/warsmash/viewer5/handlers/w3x/simulation/behaviors/build/CBehaviorHumanBuild.java

```diff
@@ -1,4 +1,164 @@
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.CUnit;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.CUnitType;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.CAbility;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.build.CAbilityBuildInProgress;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.build.CAbilityHumanRepair;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.targeting.AbilityPointTarget;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.behaviors.CAbstractRangedBehavior;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.behaviors.CBehavior;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.pathing.CBuildingPathingType;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.players.CPlayer;
+
+import java.awt.image.BufferedImage;
+import java.util.EnumSet;
+
+public class CBehaviorHumanBuild extends CAbstractRangedBehavior {
+    private static int delayAnimationTicks = (int) (2.267f / WarsmashConstants.SIMULATION_STEP_TIME);
+    private int highlightOrderId;
+    private War3ID orderId;
+    private boolean unitCreated = false;
+    private boolean done = false;
+    private CUnit constructedStructure =  null;
+
+    public CBehaviorHumanBuild(CUnit unit) {
+        super(unit);
+    }
+
+    public CBehavior reset(final AbilityPointTarget target, final int orderId, final int highlightOrderId) {
+        this.highlightOrderId = highlightOrderId;
+        this.orderId = new War3ID(orderId);
+        this.unitCreated = false;
+        this.done = false;
+        return innerReset(target);
+    }
+
+    @Override
+    public boolean isWithinRange(final CSimulation simulation) {
+        if (this.done != false) {
+            return true;
+        }
+        final CUnitType unitType = simulation.getUnitData().getUnitType(this.orderId);
+        final BufferedImage buildingPathingPixelMap = unitType.getBuildingPathingPixelMap();
+        if (buildingPathingPixelMap == null) {
+            return this.unit.canReach(this.target.getX(), this.target.getY(), unitType.getCollisionSize());
+        }
+        return this.unit.canReachToPathing(0, simulation.getGameplayConstants().getBuildingAngle(),
+                buildingPathingPixelMap, this.target.getX(), this.target.getY());
+    }
+
+    @Override
+    public int getHighlightOrderId() {
+        return this.highlightOrderId;
+    }
+
+    @Override
+    protected CBehavior update(final CSimulation simulation, final boolean withinFacingWindow) {
+        if (this.done) {
+                if(this.constructedStructure==null){
+                    return this.unit.pollNextOrderBehavior(simulation);
+                }else{
+                    CBehaviorHumanRepair repair = new CBehaviorHumanRepair(this.unit,this.unit.getFirstAbilityOfType(CAbilityHumanRepair.class));
```

Comment:
Maybe there could be a slight performance benefit here if we return `unit.getFirstAbilityOfType(...).begin(...)` instead of allocating a new behavior? In theory the ability probably already has a behavior that its recycling over its lifecycle. Should work fine either way though, just saves a memory allocation.

The "experimental" branch version of this also calls `ability.checkCanUse(...)` and `ability.checkCanTarget(...)` on the repair ability before calling `ability.begin(...)` on it. In theory, I guess if we had a peasant who was allowed to build stuff but not allowed to repair it because he could only repair organic ancients or something, maybe that would work better if we check targeting like that?

Maybe it's not relevant, but I had a Warcraft III map as a kid with a crazy worker who could start building human structures but didn't have repair and wasn't allowed to finish them. I think it was mostly just a test to play around with the World Editor, though.